File: PythonQuestions/python_questions/hard.py

""" Print the disassembly of the function object "test"
    defined below.
"""
def question1():
    import dis as d
    def test():
        return {'a': 1, 'b': 2}
    #print disassembly here, you may wish to import items in the stdlib if necessary
    d.dis(test)
    

""" Using the predefined list, "test_list", group the list
    into new lists by the first letter of each item in the list.
    Your final solution should return a list of lists whose members
    all have the same first letter.
"""
def question2():
    test_list = ["alligator", "bear", "cat", "crocodile", "beaver", "ape", "bee", "coati", "aardvark"]
    grouped_list = []
    # group the list by first letter here, you may wish to import items in the stdlib if necessary
    intermediate = {}
    for word in test_list: 
        if word[0] in intermediate.keys():
            intermediate[word[0]].append(word)
        else:
            intermediate[word[0]] = [word]
    [grouped_list.append(x) for x in intermediate.values()]
    print grouped_list
    return grouped_list

""" Create a class, "TinyObject", which uses up a very small amount of memory per instance.
    The class must have 3 instance attributes, "a", "b", and "c", and a readonly instance property,
    "combo" which simply returns the concatenated values of "a", "b", and "c"
"""
def question3():
    TinyObject = None
    # create the class here, you may wish to import items in the stdlib if necessary
    return TinyObject

question2()
question1()

File: JavascriptQuestions/hard.js
/* 
 * Argument handling
 * Write a function that takes an arbirary number of numeric arguments
 * and return the sum of those arguments
 */
function question1()
{
	var sum = null;
	// define the sum function here
	
	return sum;
}

/*
 * Using the predefined list, "test_list", group the list
 * into new lists by the first letter of each item in the list.
 * Your final solution should return a list of lists whose members
 * all have the same first letter.
 */
function question2()
{
	var test_list = ['alligator', 'bear', 'cat', 'crocodile', 'beaver', 'ape', 'bee', 'coati', 'aardvark'];
	var grouped_list = [];
	var groups = {};
	// group the list by first letter here
	for ( var i = 0; i < test_list.length; i ++ )
	{
		var item      = test_list[ i ];
		var firstChar = item.charAt( 0 );
		if(groups[firstChar]!=undefined){
			groups[ firstChar ].push( item );
		}else{
			groups[firstChar] = [];
		}
	}
	for(var key in groups) {
	    if(groups.hasOwnProperty(key)) {
	        grouped_list.push(groups[key].sort());
	    }
	}
	console.log(grouped_list);
   	return grouped_list;
} 

/*
 * map & reduce
 * In text processing, letters and combinations of letters are
 * often analyzed using numerical interpretations. Given a rule for
 * calculating the numerical "total value" of a string, wherein each letter of the string is worth
 * its ascii value (i.e. letter.charCodeAt(0) for each letter) and then summed, use map & reduce
 * methodology to calculate the "total value" of each string in the array "names", sum the "total values",
 * and then calculate the average.
 */
 function question3()
 {
	var averageTotalValue = null;
	var names = ["David", "Jimmy", "Matt", "Andy", "Nick", "Nathan", "Andrea", "Stephen", "Justin", "Emily", "Hannah"];
	var mapCallback = function(element, index, context) {
		var total = 0;
		for(var i=0; i<element.length;i++){
			total+=(element[i].charCodeAt(0) - 64)
		}
		return total;
	};
	var reduceCallback = function(previousValue, currentValue, index, context) { 
		averageTotalValue +=currentValue;
	};
	//call map and reduce on the array to calculate the sum of the words
    names.map(mapCallback).reduce(reduceCallback,0);
    //the average is the total divided by the number of items.
	averageTotalValue = (averageTotalValue/names.length);
	return averageTotalValue;
 }
